### PR TITLE
[HttpKernel] Fix nullable types handling

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q | A |
| --- | --- |
| Branch? | 3.1 |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #19784 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
The current code builds on top of a non existing semantical difference between `isNullable` and `allowsNull`. The engine has only one single state for these concepts. As such, we can't build something reliable on side effects of the new notation for nullable types as introduced in PHP 7.1. See e.g. [this comment](https://github.com/php/php-src/pull/2137#issuecomment-248940853) on the topic.
This PR makes both "concepts" collapse into a single one.

**Comment by nicolas-grekas** [comment]
Fixes the current failures on 3.1 for HHVM.
Ping @iltar FYI

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/DefaultValueResolver.php [hunk 1]
```diff
@@ -27,7 +27,7 @@
      */
     public function supports(Request $request, ArgumentMetadata $argument)
     {
-        return $argument->hasDefaultValue() || $argument->isNullable();
+        return $argument->hasDefaultValue() || ($argument->isNullable() && !$argument->isVariadic());
```
Comment (iltar) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/DefaultValueResolver.php]: I don't think variadic works here either way (php 7.0 and 7.1):
```
php > function (...$foo = []) {};
PHP Fatal error:  Variadic parameter cannot have a default value in php shell code on line 1
```

**Review by linaori — APPROVED** [review]

File: src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/DefaultValueResolver.php (hunk 1 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/DefaultValueResolver.php]: but variadics allow null: `function foo(...$foo) {};` => `foo(null)` is allowed, that's the point here.

Comment (iltar) [inline src/Symfony/Component/HttpKernel/Controller/ArgumentResolver/DefaultValueResolver.php]: Ah the previous case wouldn't even reach here, fair enough

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Pull request merged**